File: ADS_2_POO/Python Version/documentos.py
from typing import List


class Documentos:
    def __init__(self, numero, tamanho_com_digitos, tamanho_sem_digitos):
        self.numero = numero
        self.tamanho_com_digitos = tamanho_com_digitos
        self.tamanho_sem_digitos = tamanho_sem_digitos
        self.numero_correto = "a"
        self.mascara_numero = "a"
        self.peso_um = [0, 1]
        self.peso_final = [0, 1]

    def calcular_digito(self, numero, pesos):
        soma = sum(int(numero[i]) * pesos[i] for i in range(min(len(numero), len(pesos))))
        resto = soma % 11
        resultado = 11 - resto
        return '0' if resultado > 9 else str(resultado)

    def calcular_numero_correto(self, numero):
        self.numero_correto = numero[:self.tamanho_sem_digitos] + self.calcular_digito(numero, self.peso_um)
        self.numero_correto = self.numero_correto + self.calcular_digito(self.numero_correto, self.peso_final)
        return self.numero_correto

    def validar(self):
        resposta = f"{self.mascara_numero} {self.numero_correto} "
        if len(self.numero) == self.tamanho_com_digitos - 2:
            self.numero = self.numero + self.calcular_digito(self.numero, self.peso_um)

        if len(self.numero) == self.tamanho_com_digitos - 1:
            self.numero = self.numero + self.calcular_digito(self.numero, self.peso_final)

        resposta = resposta + ("Valido" if self.numero_correto == self.numero else "Invalido")
        return resposta

File: ADS_2_POO/ADS2/Documentos.cs
using System.Text.RegularExpressions;
using static DocumentosFormatacao;

public abstract class Documentos
{
    protected string numero;
    protected string numeroCorreto = "a";
    protected string mascaraNumero = "a";
    protected int[] pesoUm = { 0, 1 };
    protected int[] pesoFinal = { 0, 1};
    protected int tamanhoComDigitos;
    protected int tamanhoSemDigitos;
    

    public Documentos(string numero, int tamanhoComDigitos, int tamanhoSemDigitos)
    {
        this.numero = numero;
        this.tamanhoComDigitos = tamanhoComDigitos;
        this.tamanhoSemDigitos = tamanhoSemDigitos;
    }


    public virtual string CalcularDigito(string numero, int[] pesos)
    {
        int soma = 0;
        int length = Math.Min(numero.Length, pesos.Length);//Por que pesos.Lenght - 1 n funcionou?

        for (int i = 0; i < length; i++)
        {
            soma += int.Parse(numero[i].ToString()) * pesos[i];
        }

        int resto = soma % 11;
        int resultado = 11 - resto;
        return ((resultado > 9) ? 0 : resultado).ToString();

    }

    public string CalcularNumeroCorreto(string numero)  
    {
        numeroCorreto = (numero[..(tamanhoSemDigitos)] + CalcularDigito(numero, pesoUm));
        numeroCorreto = (numeroCorreto + CalcularDigito(numeroCorreto, pesoFinal));
        return numeroCorreto;
    }

    public virtual string Validar()
    {
       string resposta = mascaraNumero + " " + numeroCorreto + " ";
       if (numero.Length == tamanhoComDigitos - 2)
       {
            numero = (numero + CalcularDigito(numero, pesoUm));
       }

       if (numero.Length == tamanhoComDigitos - 1)
       {
            numero = (numero + CalcularDigito(numero, pesoFinal));
       }

        resposta = resposta + ((numeroCorreto == numero) ? "Valido" : "Invalido");
        return resposta;
    }
}
